tests/unit/test_exchange_format.py

```diff
@@ -0,0 +1,166 @@
+"""Tests for importing and exporting USDB IDs"""
+
+import os
+
+from usdb_syncer import SongId
+from usdb_syncer.exchange_format import USDBIDFileParser
+
+KEY_IDS = "ids"
+KEY_ERRORS = "errors"
+
+ids_per_file: dict[str, dict] = {
+    "broken_format.desktop": {
+        KEY_IDS: [],
+        KEY_ERRORS: ["invalid file format: missing a section header"],
```

Comment:
good idea, i will change that. Indeed I was not happy with this solution

Reply:
addressed in 884ee4e
- I separated valid from invalid tests as valid tests always had empty errors and invalid tests always had empty song_ids
- invalid tests forcus on exception type
- For most tests the type itself is explicit enough, these test for message contains empty string `""` (=always true), __but__ in order to provide useful/debuggable errors to the user I still would like to test for certain information provided within the error in some cases - Is this ok?

Reply:
Much better, thanks!
If you want to test context information you should store it in the error instance instead of constructing the error message directly. Python exceptions usually define the message in `__str__`.
Let me give a simplified example:

```py
@attrs.define
class UsdbUrlParseError(UsdbParseError):
    """Raised when a URL does not contain a USDB Id."""
    bad_url: str

    def __str__(self) -> str:
        return f"Failed to parse song id from URL '{self.bad_url}'"

def parse_url(url: str) -> SongId:
    ...
    raise UsdbUrlParseError(url)

# handling the error
try:
    url = parse_url("https://github.com/bohning/usdb_syncer/pull/95")
except UsdbParseError as error:
    print(error)

# testing the error
with pytest.raises(UsdbUrlParseError) as exc_info:
    parse_url("https://github.com/bohning/usdb_syncer/pull/95")

assert exc_info.value.bad_url == "https://github.com/bohning/usdb_syncer/pull/95"
```

Hope that made things clearer!

Reply:
Ah thanks for the example, I think this might also help with my problems with @attes.define above

Reply:
addressed in 5806e4b